File: python/tournament.py
class Team:

    def __init__(self, name):
        self.name = name
        self.mp = 0
        self.w = 0
        self.d = 0
        self.l = 0

    def points(self):
        return self.w * 3 + self.d

    def win(self):
        self.mp += 1
        self.w += 1

    def loss(self):
        self.mp += 1
        self.l += 1

    def draw(self):
        self.mp += 1
        self.d += 1

    def __str__(self):
        return f'{self.name: <31}|{self.mp: >3} |{self.w: >3} |{self.d: >3} |{self.l: >3} |{self.points(): >3}'


def tally(rows):
    teams = {}
    for row in rows:
        [team_a, team_b, result] = row.split(';')
        if team_a not in teams:
            teams[team_a] = Team(team_a)
        if team_b not in teams:
            teams[team_b] = Team(team_b)

        if result == 'win':
            teams[team_a].win()
            teams[team_b].loss()
        elif result == 'draw':
            teams[team_a].draw()
            teams[team_b].draw()
        else:
            teams[team_a].loss()
            teams[team_b].win()

    table = ["Team                           | MP |  W |  D |  L |  P"]
    team_lst = sorted(teams.values(), key=lambda t: (-t.points(), t.name))
    for team in team_lst:
        table.append(f'{team}')
    return table

File: elixir/tournament.ex
defmodule Tournament do
  @header "Team                           | MP |  W |  D |  L |  P"

  @doc """
  Given `input` lines representing two teams and whether the first of them won,
  lost, or reached a draw, separated by semicolons, calculate the statistics
  for each team's number of games played, won, drawn, lost, and total points
  for the season, and return a nicely-formatted string table.

  A win earns a team 3 points, a draw earns 1 point, and a loss earns nothing.

  Order the outcome by most total points for the season, and settle ties by
  listing the teams in alphabetical order.
  """
  @spec tally(input :: list(String.t())) :: String.t()
  def tally(input) do
    input
    |> Enum.filter(fn s ->
      s =~ ~r/^.*;.*;(win|draw|loss)$/
    end)
    |> Enum.reduce(%{}, fn s, acc ->
      [a, b, res] = String.split(s, ";")

      case res do
        "win" -> {[1, 1, 0, 0], [1, 0, 0, 1]}
        "loss" -> {[1, 0, 0, 1], [1, 1, 0, 0]}
        "draw" -> {[1, 0, 1, 0], [1, 0, 1, 0]}
      end
      |> then(fn {r1, r2} ->
        acc
        |> Map.update(a, r1, &add(&1, r1))
        |> Map.update(b, r2, &add(&1, r2))
      end)
    end)
    |> Enum.sort_by(fn {k, v} -> {-calc(v), k} end)
    |> Enum.map(fn {k, v} ->
      [mp, w, d, l, p] =
        (v ++ [calc(v)])
        |> Enum.map(fn n ->
          String.pad_leading("#{n}", 3)
        end)

      "#{String.pad_trailing(k, 31)}|#{mp} |#{w} |#{d} |#{l} |#{p}"
    end)
    |> then(&[@header | &1])
    |> Enum.join("\n")
  end

  defp add(xs, ys) do
    Enum.zip(xs, ys)
    |> Enum.map(fn {x, y} -> x + y end)
  end

  defp calc([_mp, w, d, _l]) do
    w * 3 + d * 1
  end
end
